### PR TITLE
added embedded model support, cleaned up error structure a bit

### PR DESCRIPTION
**Pull request by SimonSchick**
Accidently pushed this to master, hence this is a revert of revert.

### CODE REVIEW

**Comment by SimonSchick** [comment]
@connor4312 @JamyDev for review since review bot is derp

File: src/errors/ReadError.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { IModel } from '../Types';
+import { BormError } from './BormError';
+
+/**
+ * Error for
```
Comment (JamyDev) [inline src/errors/ReadError.ts]: Error for what? Potatoes? :mappa:

File: src/errors/ValidationError.ts [hunk 2]
```diff
@@ -1,14 +1,13 @@
+import { IValidationResult } from '../Types';
 import { BormError } from './BormError';
-import { IValidationResult } from './Types';
 
 /**
  * Base error that is used as a base for all errors that can occur when running requests.
```
Comment (JamyDev) [inline src/errors/ValidationError.ts]: Please change this to:

`Error class used for validation errors returned by requests`

File: src/errors/WriteError.ts [hunk 3]
```diff
@@ -1,13 +1,12 @@
+import { IColumnFactory } from '../Types';
 import { BormError } from './BormError';
-import { IColumnFactory } from './Types';
 
 /**
  * Base error that is used as a base for all errors that can occur when running requests.
```
Comment (JamyDev) [inline src/errors/WriteError.ts]: `Error that is used when a consumer tries to write to a read-only property`

File: src/errors/ValidationError.ts (hunk 2 above)
Comment (JamyDev) [inline src/errors/ValidationError.ts]: Thanks!

File: readme.md [hunk 4]
```diff
@@ -269,3 +269,27 @@ myAdapter.createResource(myUserInstance.avatar, imageDataBlobl)
...
+```typescript
+class User extends Model {
+    public static slug = 'user';
+
+    @Column({ primaryKey: true })
+    public id: IColumn<number>;
+
+    @Embed({
+        model: Project,
+        multi: true,
```
Comment (connor4312) [inline readme.md]: Hm, any way you can you pick this up from the metadata?

File: src/Model.ts [hunk 5]
```diff
@@ -150,6 +155,31 @@ export abstract class Model implements IModel {
...
+                    this[key] = Observable.never();
+                    return;
+                }
+                if (embedOptions.multi && !(val instanceof Array)) {
+                    throw new ReadError(`Expected embedded model property ${key} to be an array but got single value`, this, key, val);
+                }
+                if (!embedOptions.multi && val instanceof Array) {
+                    throw new ReadError(`Expected embedded model property ${key} to be value but got an array`, this, key, val);
+                }
+                const ctor = (<IModelCtor<IModel>>embedOptions.model).slug ?
```
Comment (connor4312) [inline src/Model.ts]: You could do `'slug' in embedOptions.model` to avoid a bit of type gore

File: src/Model.ts [hunk 6]
```diff
@@ -150,6 +155,31 @@ export abstract class Model implements IModel {
...
+                    <IModelCtor<IModel>>embedOptions.model :
+                    (<() => IModelCtor<IModel>>embedOptions.model)();
+                if (embedOptions.multi) {
+                    this[key] = new BehaviorSubject((<Object[]>val).map(subRawModel => new ctor(subRawModel, options)));
+                } else {
+                    this[key] = new BehaviorSubject(new ctor(val, options));
+                }
+                delete values[key];
```
Comment (connor4312) [inline src/Model.ts]: Imo we should not mutate the provided values

**Comment by connor4312** [comment]
It would be super great if we could short-circuit `model.embedded._subscribe` to request `model/{id}/embeddedSlug` if there isn't already a value for it.

File: src/Model.ts (hunk 5 above)
Comment (SimonSchick) [inline src/Model.ts]: I rewrote most of this in the partial-refactor, I'll let that sit for now as I don't want to introduce too many conflicts.

**Comment by SimonSchick** [comment]
Problem with that is not all embedded models are associated in our api, eg a channel may have their user embedded and vice versa but there is no `/channels/:id/user` or `/users/:id/channel`, this is more what associations are for.

I initially wanted to combine these but our API has a few quirks that don't quite allow a clean implementation of combined embed/associations unfortunately 😞

File: src/Model.ts (hunk 6 above)
Comment (SimonSchick) [inline src/Model.ts]: yea, good point here, I hope we soon get support for `const` parameters (like in C++) so you can't alter a value in any way.

Comment (SimonSchick) [inline src/Model.ts]: You actually made me aware of a flaw in my logic, fixing it.

Comment (SimonSchick) [inline src/Model.ts]: Fixed it.

**Pull request merged**